src/kakarot/instructions/system_operations.cairo

```diff
@@ -301,6 +301,60 @@ namespace SystemOperations {
 
         return sub_ctx;
     }
+
+    // @notice SELFDESTRUCT operation.
+    // @dev
+    // @custom:since Frontier
+    // @custom:group System Operations
+    // @custom:gas 3000 + dynamic gas
+    // @custom:stack_consumed_elements 1
+    // @return The pointer to the updated execution_context.
+    func exec_selfdestruct{
+        syscall_ptr: felt*,
+        pedersen_ptr: HashBuiltin*,
+        range_check_ptr,
+        bitwise_ptr: BitwiseBuiltin*,
+    }(ctx: model.ExecutionContext*) -> model.ExecutionContext* {
+        alloc_locals;
+
+        // Get stack and memory from context
+        let stack = ctx.stack;
+
+        // Stack input:
+        // 0 - address: account to send the current balance to
+        let (stack, address_uint256) = Stack.pop(stack);
+        let address_felt = Helpers.uint256_to_felt(address_uint256);
+        let ctx = ExecutionContext.update_stack(ctx, stack);
+        // Get the number of native tokens owned by the given starknet
+        // account and transfer them to receiver
+        let (native_token_address_) = native_token_address.read();
+        let (balance: Uint256) = IEth.balanceOf(
+            contract_address=native_token_address_, account=ctx.starknet_contract_address
+        );
+        let (success) = IEth.transfer(contract_address=native_token_address_, recipient=address_felt, amount=balance);
```

Comment:
Do you know how errors are managed on these instructions? For example if there is an issue with the selfdestruct later in the code, would that revert the transaction at a whole? I think we should transfer the token after the contract has been destructed instead of before to not credit the client with contract balance if there are issues with selfdestruction

Reply:
From my understanding balance is transferred immediately when selfdestruct is called but contract destroyed only at the end of the tx. I don't have any example coming to my mind but transferring balance only at the end of the tx  may bring unwanted behaviors. Also if tx fails everything is reverted (no contract destruction and no balance transfer). So if anything fail in selfdestruct, transfer would also be reverted. Or am I missing something ?